src/core/logging/logger.hpp

```diff
@@ -324,14 +329,16 @@
 
 #endif
 
-#define log_and_throw(message)                                      \
-  do {                                                              \
-    auto throw_error = [&]() GL_COLD_NOINLINE_ERROR {  \
-      logstream(LOG_ERROR) << (message) << std::endl;               \
-      throw(std::string(message));                                  \
-    };                                                              \
-    throw_error();                                                  \
-  } while(0)
+#ifdef NDEBUG
+
+#define log_and_throw(message)                        \
+  do {                                                \
+    auto throw_error = [&]() GL_COLD_NOINLINE_ERROR { \
+      logstream(LOG_ERROR) << (message) << std::endl; \
+      throw(std::runtime_error(message);              \
```

Comment:
I forget to mention one more important point: 

Change the default `log_and_throw` behavior. Previous behavior is to `throw(std::string(message))`, which is really hard to catch since it's not an `std::exception` type and you need to specify error handling logic like `catch(const string& ex)`. That's why there're lots of `catch(...)` around the codebase, which is a bad practice since you miss the chance to interpret the exception message or propagate the original exception message. After we set `runtime_error` as default exception, we can use `catch (const std::exception& ex)` to avoid `catch(...)`.

Meanwhile, we also need to restrict the exception categories.  Unlimited different types of expcetions will make you lazy and tempt you to use `catch (const std::exception& ex)` as a brutal solution. Even though it can help, but we should have provided error handling for certain recoverable exceptions. 

Therefore, I think we need to use only few common exception types like `runtime`, `io` and `engine_optimizer` specific exception. Those types of exceptions should be the only types of exceptions we expect. In this way, we can expect a fairly hierarchical  error-handling or logging to react to specific errors.
```cpp
try {
} catch (const std::runtime_error & ex) {
} catch (const std::turi::fileio::io_error& ex) {
} catch (const std::exception& ex) {
} catch (...) {
// unknown which should not happen
}
```

Reply:
@ylow -- do you think this is okay to do in general?

Reply:
or with a more general io_failiure exception handling.

Reply:
I think it's ok to do in general, but I would prefer to see a separate PR for it.